blog/02-18-2023-Build-an-NFT-Full-Stack-Monster-War-Game-on-The-Celo-Blockchain/Build-an-NFT-Full-Stack-Monster-War-Game-on-The-Celo-Blockchain.md

```diff
@@ -0,0 +1,998 @@
+
+```solidity
+    mapping(address => bool) public minters;
+    mapping (uint => NFT) internal nfts;
+    mapping (address => uint) public playerpowervalue;
+```
+
+We also declare three mappings. first, `minters` (a mapping addresses to booleans that represents whether an address is a minter or not), "nfts" (a mapping uint to NFT to accomodate the list of nfts minted) and `playerpowervalue` (a mapping address to uint to keep track of the power value of each player).
+
+```solidity
+  modifier hasmint(address _address) {
+        require(minters[_address], "Invalid address");
+        _;
+    }
+
+     modifier canSwallow(address _address, uint _index) {
+       address ownerAddress = nfts[_index].owner;
+        require(playerpowervalue[_address] > playerpowervalue[ownerAddress], "You have less value point");
+        _;
+    }
+```
+
+Next we added two modifiers. 
+
+`hasmint` This modifier checks whether the address calling the function has already minted an NFT by verifying that the address is a valid minter. If the address is not a valid minter, the function execution is aborted, otherwise it is allowed to proceed.
+
+`canSwallow` This modifier checks whether the power value of the attacker is greater than the power value of the owner of the NFT before allowing the `swallowNFT()` function to execute. If the power value of the attacker is less than or equal to the owner's power value, the function execution is aborted.
+
+let's look at the functions
+
+```solidity
+ function mint( string memory name) public payable {
+        uint256 tokenId = _tokenIdCounter.current();
+        _tokenIdCounter.increment();
+        _safeMint(msg.sender, tokenId);
+        addNFT(tokenId, name);// listing the nft 
+    }
+```
+
+Let's declare a function called `mint()` The Mint function allows a player to mint their own NFTs. The function takes a string as a parameter which is used to specify the name of the NFT. The function first generates a unique `tokenId` for the NFT and increments the `tokenId counter`. It then calls the `_safeMint` function which is inherited from the `ERC721` contract and is used to mint the NFT. Finally, it calls the `addNFT` function which adds the NFT to the game.
+
+```solidity
+ function addNFT(uint256 _tokenId, string memory _name) private{
+        uint _powerValue = 0;
+        nfts[allNFTs] = NFT(
+            _tokenId,
+            payable(msg.sender),
+            _powerValue,
+            _name
+        );
+        allNFTs++;
+        minters[msg.sender] = true;
+    }
+```
+
+The `AddNFT()` function adds the NFT to the game. The function takes a `tokenId` and a name as parameters. It then creates a new NFT struct with the given parameters and adds it to the internal nfts mapping. It also adds the player to the minters mapping and increments the `allNFTs` variable. 
```

Comment:
The `AddNFT()` function adds the NFT to the game. The function takes a `tokenId` and a name as parameters. It then creates a new NFT struct with the given parameters and adds it to the internal NFTs mapping. It also adds the player to the minters mapping and increments the `allNFTs` variable.